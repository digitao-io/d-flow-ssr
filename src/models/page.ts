export interface Page {
  key: string;
  title: string;
  description: string;
  urlPattern: string;

  details: PageDetails;
}
export interface PageDetails {
  template: string;
  language: ConfigValue;
  title: ConfigValue;
  config: Record<string, ConfigValue>;
  slots: Record<string, ComponentDetails[]>;
}

export interface ComponentDetails {
  component: string;
  config: Record<string, ConfigValue>;
}

export type ResolvedPageDetails = {
  template: string;
  language: string;
  title: string;
  config: Record<string, ResolvedValue>;
  slots: Record<string, ResolvedComponentDetails[]>;
};

export type ResolvedComponentDetails = {
  component: string;
  config: Record<string, ResolvedValue>;
};

export type ConfigValue =
  | ValueResolvingInfo
  | ResolvedValue;

export interface ValueResolvingInfo {
  $source: Array<string | ValueResolvingInfo>;
  field?: string;
}

export type ResolvedValue =
  | string
  | number
  | boolean
  | null
  | ResolvedArrayValue
  | ResolvedObjectValue;

export type ResolvedArrayValue = ResolvedValue[];
export interface ResolvedObjectValue {
  [key: string]: ResolvedValue;
}

export function isValueResolved(v: ConfigValue): v is ResolvedValue {
  return v === null
    || typeof v !== "object"
    || !(v as ValueResolvingInfo).$source;
}
