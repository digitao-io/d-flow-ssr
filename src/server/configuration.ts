import fs from "node:fs";

export interface Configuration {
  staticPath: string;
  htmlTemplatePath: string;
  port: number;
}

export function readConfiguration<CONFIG extends Configuration>(configPath: string): CONFIG {
  if (!configPath) {
    throw Error("Invalid configuration path");
  }

  if (!fs.existsSync(configPath)) {
    throw Error(`Configuration path doesn't exist: ${configPath}`);
  }

  if (!fs.statSync(configPath).isFile()) {
    throw Error(`Configuration path isn't a file: ${configPath}`);
  }

  const configFileContent = fs.readFileSync(configPath, "utf-8");

  let config: CONFIG;
  try {
    config = JSON.parse(configFileContent);
  }
  catch {
    throw Error(`Cannot parse configuration: ${configPath}`);
  }

  return config;
}
