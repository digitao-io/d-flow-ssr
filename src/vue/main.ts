export { createVueApp } from "./vue-app";
