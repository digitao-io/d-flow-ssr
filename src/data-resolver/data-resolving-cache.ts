import { ResolvedValue } from "../models/page";

export type DataCacheOption =
  | DataCacheTimeScopedOption
  | DataCacheRequestScopedOption;

export interface DataCacheTimeScopedOption {
  type: "TIME";
  timeToLive: number;
}

export interface DataCacheRequestScopedOption {
  type: "REQUEST";
}

export type DataCacheResult =
  | { hit: true; value: ResolvedValue }
  | { hit: false };

type DataCacheEntry =
  & { value: ResolvedValue }
  & (
    | DataCacheTimeEntry
    | DataCacheRequestEntry
  );

interface DataCacheTimeEntry {
  type: "TIME";
  expireAt: number;
}

interface DataCacheRequestEntry {
  type: "REQUEST";
}

export class DataResolvingCache {
  private cachedValue: Map<string, DataCacheEntry>;

  public constructor() {
    this.cachedValue = new Map();
  }

  public clearAllCachedValues() {
    this.cachedValue.clear();
  }

  public clearRequestTypedCachedValues() {
    for (const [key, entry] of this.cachedValue.entries()) {
      if (entry.type === "REQUEST") {
        this.cachedValue.delete(key);
      }
    }
  }

  public addCachedValue(
    dataPath: string[],
    value: ResolvedValue,
    cacheOption: DataCacheOption,
  ) {
    const cacheKey = dataPath.join("/");

    switch (cacheOption.type) {
      case "TIME":
        this.cachedValue.set(cacheKey, {
          type: "TIME",
          expireAt: this.getCurrentTimeInSecs() + cacheOption.timeToLive,
          value,
        });
        break;
      case "REQUEST":
        this.cachedValue.set(cacheKey, {
          type: "REQUEST",
          value,
        });
        break;
    }
  }

  public getCachedValue(
    dataPath: string[],
  ): DataCacheResult {
    const cacheKey = dataPath.join("/");
    const cacheEntry = this.cachedValue.get(cacheKey);

    if (!cacheEntry) {
      return { hit: false };
    }

    switch (cacheEntry.type) {
      case "TIME":
        if (this.getCurrentTimeInSecs() < cacheEntry.expireAt) {
          return { hit: true, value: cacheEntry.value };
        }
        else {
          this.cachedValue.delete(cacheKey);
          return { hit: false };
        }
      case "REQUEST":
        return { hit: true, value: cacheEntry.value };
      default:
        return { hit: false };
    }
  }

  private getCurrentTimeInSecs() {
    return Math.floor(Date.now() / 1000);
  }
}
