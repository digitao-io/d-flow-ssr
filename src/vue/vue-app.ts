import { App, Component, createSSRApp, defineComponent, h, useSSRContext, VNode } from "vue";
import { ResolvedPageDetails } from "../models/page";

export function createVueApp(components: Record<string, Component>): App {
  return createSSRApp(defineComponent({
    setup() {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const ctx = useSSRContext() ?? (window as any).context;

      const pageDetails = ctx.pageDetails as ResolvedPageDetails;

      function render(slots: Record<string, () => VNode[]>): VNode {
        return h(
          components[pageDetails.template],
          { config: pageDetails.config },
          slots,
        );
      };

      function renderSlots(): Record<string, () => VNode[]> {
        const result = {} as Record<string, () => VNode[]>;
        for (const [slotName, slotComponents] of Object.entries(pageDetails.slots)) {
          result[slotName] = () => slotComponents.map((component) => {
            return h(
              components[component.component],
              { config: component.config },
            );
          });
        }

        return result;
      };

      return () => render(renderSlots());
    },
  }));
}
