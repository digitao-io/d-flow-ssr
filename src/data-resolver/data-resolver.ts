import { ConfigValue, isValueResolved, PageDetails, ResolvedComponentDetails, ResolvedPageDetails, ResolvedValue, ValueResolvingInfo } from "../models/page";
import { DataCacheOption, DataResolvingCache } from "./data-resolving-cache";

export interface DataResolvingContext {
  apiBaseUrl: string;
  pageBaseUrl: string;
  pageUrlPath: string;
  pageUrlParams: Record<string, string>;
  pageUrlQueries: Record<string, string | string[]>;
};

export interface DataResolverConfig {
  prefix: string[];
  cacheOption?: DataCacheOption;
  resolve(
    context: DataResolvingContext,
    sourcePath: string[],
  ): Promise<ResolvedValue> | ResolvedValue;
}

// eslint-disable-next-line @typescript-eslint/no-empty-object-type
export interface DataResolverConfigLookup
  extends Record<string, DataResolverConfig | DataResolverConfigLookup> {}

function isDataResolverConfig(
  v: DataResolverConfig | DataResolverConfigLookup,
): v is DataResolverConfig {
  const keys = Object.keys(v);
  return keys.includes("prefix") && keys.includes("resolve");
}

export class DataResolver {
  private resolverConfigs: DataResolverConfigLookup;
  private cache: DataResolvingCache;

  public constructor() {
    this.resolverConfigs = {};
    this.cache = new DataResolvingCache();
  }

  public withResolver(resolverConfig: DataResolverConfig): this {
    let currentLookup: DataResolverConfigLookup = this.resolverConfigs;
    for (let i = 0; i < resolverConfig.prefix.length - 1; i++) {
      const lookupKey = resolverConfig.prefix[i];

      if (!currentLookup[lookupKey]) {
        currentLookup[lookupKey] = {};
      }

      currentLookup = currentLookup[lookupKey] as DataResolverConfigLookup;

      if (isDataResolverConfig(currentLookup)) {
        throw new Error(
          `Cannot register resolver ${resolverConfig.prefix.join("/")}, `
          + `because the resolver ${resolverConfig.prefix.slice(0, i + 1).join("/")} already exists`,
        );
      }
    }

    currentLookup[resolverConfig.prefix[resolverConfig.prefix.length - 1]] = resolverConfig;

    return this;
  }

  public async resolve(
    resolvingContext: DataResolvingContext,
    pageDetails: PageDetails,
  ): Promise<ResolvedPageDetails> {
    const resolvedTopLevelFields = await this.resolveRecord(
      resolvingContext,
      {
        language: pageDetails.language,
        title: pageDetails.title,
      },
    );

    const resolvedConfig = await this.resolveRecord(resolvingContext, pageDetails.config);

    const resolvedSlots = {} as Record<string, ResolvedComponentDetails[]>;
    for (const [slotName, componentDetailsArray] of Object.entries(pageDetails.slots)) {
      const resolvedComponentDetailsArray: ResolvedComponentDetails[] = [];

      for (const componentDetails of componentDetailsArray) {
        const resolvedComponentDetails = {
          component: componentDetails.component,
          config: await this.resolveRecord(resolvingContext, componentDetails.config),
        };

        resolvedComponentDetailsArray.push(resolvedComponentDetails);
      }

      resolvedSlots[slotName] = resolvedComponentDetailsArray;
    }

    this.cache.clearRequestTypedCachedValues();

    return {
      template: pageDetails.template,
      language: resolvedTopLevelFields.language as string,
      title: resolvedTopLevelFields.title as string,
      config: resolvedConfig,
      slots: resolvedSlots,
    };
  }

  private async resolveRecord(
    resolvingContext: DataResolvingContext,
    record: Record<string, ConfigValue>,
  ): Promise<Record<keyof typeof record, ResolvedValue>> {
    const result = {} as Record<keyof typeof record, ResolvedValue>;
    for (const [key, value] of Object.entries(record)) {
      if (isValueResolved(value)) {
        result[key] = value;
      }
      else {
        result[key] = await this.resolveValueUsingResolver(resolvingContext, value);
      }
    }

    return result;
  }

  private async resolveValueUsingResolver(
    resolvingContext: DataResolvingContext,
    resolvingInfo: ValueResolvingInfo,
  ): Promise<ResolvedValue> {
    const resolvedSource = [] as string[];
    for (const source of resolvingInfo.$source) {
      if (isValueResolved(source)) {
        resolvedSource.push(source as string);
      }
      else {
        resolvedSource.push((await this.resolveValueUsingResolver(resolvingContext, source)) as string);
      }
    }

    let currentLookup: DataResolverConfig | DataResolverConfigLookup = this.resolverConfigs;
    for (const lookupKey of resolvedSource) {
      currentLookup = (currentLookup as DataResolverConfigLookup)[lookupKey];
      if (!currentLookup) {
        throw Error(`Cannot find resolver for source ${resolvingInfo.$source.join("/")}`);
      }
      if (isDataResolverConfig(currentLookup)) {
        break;
      }
    }

    if (!isDataResolverConfig(currentLookup)) {
      throw Error(`Cannot find resolver for source ${resolvingInfo.$source.join("/")}`);
    }

    const resolverConfig = currentLookup as DataResolverConfig;

    let resolvedValue: ResolvedValue;

    const cacheResult = this.cache.getCachedValue(resolvedSource);
    if (cacheResult.hit) {
      resolvedValue = cacheResult.value;
    }
    else {
      try {
        resolvedValue = await resolverConfig.resolve(resolvingContext, resolvedSource);
        if (resolverConfig.cacheOption) {
          this.cache.addCachedValue(resolvedSource, resolvedValue, resolverConfig.cacheOption);
        }
      }
      catch (err) {
        console.log(err);
        resolvedValue = null;
      }
    }

    if (resolvingInfo.field) {
      const fieldKeys = resolvingInfo.field.split(".");
      for (const fieldKey of fieldKeys) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        resolvedValue = resolvedValue && ((resolvedValue as any)[fieldKey] ?? null);
      }
    }

    return resolvedValue;
  }
}
