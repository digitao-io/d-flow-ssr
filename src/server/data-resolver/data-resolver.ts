import { ConfigValue, isValueResolved, PageDetails, ResolvedComponentDetails, ResolvedPageDetails, ResolvedValue, ValueResolvingInfo } from "../models/page";
import { DataCacheOption, DataResolvingCache } from "./data-resolving-cache";
import { Configuration } from "../app/configuration";

export interface DataResolvingContext<CONFIG extends Configuration> {
  appConfig: CONFIG;
  pageUrlPath: string;
  pageUrlParams: Record<string, string>;
  pageUrlQueries: Record<string, string | string[]>;
};

export interface DataResolverConfig<CONFIG extends Configuration> {
  prefix: string[];
  cacheOption?: DataCacheOption;
  resolve(
    context: DataResolvingContext<CONFIG>,
    sourcePath: string[],
  ): Promise<ResolvedValue> | ResolvedValue;
}

// eslint-disable-next-line @typescript-eslint/no-empty-object-type
interface DataResolverConfigLookup<CONFIG extends Configuration>
  extends Record<string, DataResolverConfig<CONFIG> | DataResolverConfigLookup<CONFIG>> {}

function isDataResolverConfig<CONFIG extends Configuration>(
  v: DataResolverConfig<CONFIG> | DataResolverConfigLookup<CONFIG>,
): v is DataResolverConfig<CONFIG> {
  const keys = Object.keys(v);
  return keys.includes("prefix") && keys.includes("resolve");
}

export class DataResolver<CONFIG extends Configuration> {
  private resolverConfigs: DataResolverConfigLookup<CONFIG>;
  private cache: DataResolvingCache;

  public constructor() {
    this.resolverConfigs = {};
    this.cache = new DataResolvingCache();
  }

  public clearCache() {
    this.cache.clearAllCachedValues();
  }

  public withResolver(resolverConfig: DataResolverConfig<CONFIG>): this {
    let currentLookup: DataResolverConfigLookup<CONFIG> = this.resolverConfigs;
    for (let i = 0; i < resolverConfig.prefix.length - 1; i++) {
      const lookupKey = resolverConfig.prefix[i];

      if (!currentLookup[lookupKey]) {
        currentLookup[lookupKey] = {};
      }

      currentLookup = currentLookup[lookupKey] as DataResolverConfigLookup<CONFIG>;

      if (isDataResolverConfig(currentLookup)) {
        throw new Error(
          `Cannot register resolver ${resolverConfig.prefix.join("/")}, `
          + `because the resolver ${resolverConfig.prefix.slice(0, i + 1).join("/")} already exists`,
        );
      }
    }

    currentLookup[resolverConfig.prefix[resolverConfig.prefix.length - 1]] = resolverConfig;

    return this;
  }

  public async resolve(
    resolvingContext: DataResolvingContext<CONFIG>,
    pageDetails: PageDetails,
  ): Promise<ResolvedPageDetails> {
    const resolvedTopLevelFields = await this.resolveRecord(
      resolvingContext,
      {
        language: pageDetails.language,
        title: pageDetails.title,
      },
    );

    const resolvedConfig = await this.resolveRecord(resolvingContext, pageDetails.config);

    const resolvedSlots = {} as Record<string, ResolvedComponentDetails[]>;
    for (const [slotName, componentDetailsArray] of Object.entries(pageDetails.slots)) {
      const resolvedComponentDetailsArray: ResolvedComponentDetails[] = [];

      for (const componentDetails of componentDetailsArray) {
        const resolvedComponentDetails = {
          component: componentDetails.component,
          config: await this.resolveRecord(resolvingContext, componentDetails.config),
        };

        resolvedComponentDetailsArray.push(resolvedComponentDetails);
      }

      resolvedSlots[slotName] = resolvedComponentDetailsArray;
    }

    this.cache.clearRequestTypedCachedValues();

    return {
      template: pageDetails.template,
      language: resolvedTopLevelFields.language as string,
      title: resolvedTopLevelFields.title as string,
      config: resolvedConfig,
      slots: resolvedSlots,
    };
  }

  private async resolveRecord(
    resolvingContext: DataResolvingContext<CONFIG>,
    record: Record<string, ConfigValue>,
  ): Promise<Record<keyof typeof record, ResolvedValue>> {
    const result = {} as Record<keyof typeof record, ResolvedValue>;
    for (const [key, value] of Object.entries(record)) {
      if (isValueResolved(value)) {
        result[key] = value;
      }
      else {
        result[key] = await this.resolveValueUsingResolver(resolvingContext, value);
      }
    }

    return result;
  }

  private async resolveValueUsingResolver(
    resolvingContext: DataResolvingContext<CONFIG>,
    resolvingInfo: ValueResolvingInfo,
  ): Promise<ResolvedValue> {
    const resolvedSource = [] as string[];
    for (const source of resolvingInfo.$source) {
      if (isValueResolved(source)) {
        resolvedSource.push(source as string);
      }
      else {
        resolvedSource.push((await this.resolveValueUsingResolver(resolvingContext, source)) as string);
      }
    }

    let currentLookup: DataResolverConfig<CONFIG> | DataResolverConfigLookup<CONFIG> = this.resolverConfigs;
    for (const lookupKey of resolvedSource) {
      currentLookup = (currentLookup as DataResolverConfigLookup<CONFIG>)[lookupKey];
      if (!currentLookup) {
        throw Error(`Cannot find resolver for source ${resolvingInfo.$source.join("/")}`);
      }
      if (isDataResolverConfig(currentLookup)) {
        break;
      }
    }

    if (!isDataResolverConfig(currentLookup)) {
      throw Error(`Cannot find resolver for source ${resolvingInfo.$source.join("/")}`);
    }

    const resolverConfig = currentLookup as DataResolverConfig<CONFIG>;

    let resolvedValue: ResolvedValue;

    const cacheResult = this.cache.getCachedValue(resolvedSource);
    if (cacheResult.hit) {
      resolvedValue = cacheResult.value;
    }
    else {
      try {
        resolvedValue = await resolverConfig.resolve(resolvingContext, resolvedSource);
        if (resolverConfig.cacheOption) {
          this.cache.addCachedValue(resolvedSource, resolvedValue, resolverConfig.cacheOption);
        }
      }
      catch (err) {
        console.log(err);
        resolvedValue = null;
      }
    }

    if (resolvingInfo.field) {
      const fieldKeys = resolvingInfo.field.split(".");
      for (const fieldKey of fieldKeys) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        resolvedValue = resolvedValue && ((resolvedValue as any)[fieldKey] ?? null);
      }
    }

    return resolvedValue;
  }
}
