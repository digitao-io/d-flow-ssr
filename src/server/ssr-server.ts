import fs from "node:fs";
import express from "express";
import { Express } from "express";
import sirv from "sirv";
import helmet from "helmet";
import compression from "compression";
import { App } from "vue";
import { DataResolver } from "../data-resolver/data-resolver";
import { DynamicRouter } from "./dynamic-router";
import { Configuration, readConfiguration } from "./configuration";
import { Page } from "../models/page";

export interface SsrServerConfig<CONFIG extends Configuration> {
  vueApp: App;
  dataResolver: DataResolver<CONFIG>;
  fetchPages: PagesProvider;
  configPath: string;
}

export type PagesProvider = () => Promise<Page[]>;

export class SsrServer<CONFIG extends Configuration> {
  private dynamicRouter: DynamicRouter<CONFIG>;
  private dataResolver: DataResolver<CONFIG>;
  private fetchPages: PagesProvider;
  private htmlTemplate: string;
  private config: CONFIG;

  public express: Express;

  public constructor() {
    this.dynamicRouter = null as unknown as DynamicRouter<CONFIG>;
    this.dataResolver = null as unknown as DataResolver<CONFIG>;
    this.fetchPages = null as unknown as PagesProvider;
    this.htmlTemplate = null as unknown as string;
    this.config = null as unknown as CONFIG;

    this.express = null as unknown as Express;
  }

  public async initialize(config: SsrServerConfig<CONFIG>) {
    this.config = readConfiguration(config.configPath);
    this.htmlTemplate = fs.readFileSync(this.config.htmlTemplatePath, "utf8");

    this.dynamicRouter = new DynamicRouter(
      config.vueApp,
      config.dataResolver,
      this.htmlTemplate,
      this.config,
    );
    this.dataResolver = config.dataResolver;
    this.fetchPages = config.fetchPages;

    const pages = await this.fetchPages();
    this.dynamicRouter.buildRoutes(pages);

    this.express = express();

    this.express.use(helmet({
      contentSecurityPolicy: {
        directives: {
          "script-src": ["'self'", "'unsafe-inline'"],
          "img-src": ["'self'", "data:", "https:"],
        },
      },
    }));

    this.express.use(compression());

    this.express.use("/", sirv(this.config.staticPath, { extensions: [] }));

    this.express.post("/maintenance/cache/clear", (_, res) => {
      this.dataResolver.clearCache();

      res.status(200);
      res.json({ status: "OK" });
      res.end();
    });

    this.express.post("/maintenance/page/fetch", async (_, res) => {
      const pages = await this.fetchPages();
      this.dynamicRouter.buildRoutes(pages);

      res.status(200);
      res.json({ status: "OK" });
      res.end();
    });

    this.express.use("/", (req, res, next) => {
      this.dynamicRouter.router(req, res, next);
    });
  }

  public run() {
    this.express.listen(this.config.port, () => {
      console.log(`Server started at port ${this.config.port}`);
    });
  }
}
