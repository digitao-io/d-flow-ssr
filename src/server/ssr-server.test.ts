import supertest from "supertest";
import { App, Component, createSSRApp, defineComponent, h, useSSRContext, VNode } from "vue";
import { SsrServer } from "./ssr-server";
import { Configuration } from "./configuration";
import { DataResolver } from "../data-resolver/data-resolver";
import { ResolvedPageDetails } from "../models/page";

export function createVueApp(components: Record<string, Component>): App {
  return createSSRApp(defineComponent({
    setup() {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const ctx = useSSRContext() ?? (window as any).context;

      const pageDetails = ctx.pageDetails as ResolvedPageDetails;

      function render(slots: Record<string, () => VNode[]>): VNode {
        return h(
          components[pageDetails.template],
          { config: pageDetails.config },
          slots,
        );
      };

      function renderSlots(): Record<string, () => VNode[]> {
        const result = {} as Record<string, () => VNode[]>;
        for (const [slotName, slotComponents] of Object.entries(pageDetails.slots)) {
          result[slotName] = () => slotComponents.map((component) => {
            return h(
              components[component.component],
              { config: component.config },
            );
          });
        }

        return result;
      };

      return () => render(renderSlots());
    },
  }));
}

async function initializeSsrServer() {
  const vueApp = createVueApp({
    "default-template": {
      props: {
        config: { type: Object, required: true },
      },
      template: `
        <main>
          <h1>Hello world!</h1>
          <p>message: {{ config.message }}</p>
          <slot name="main" />
          <slot name="sidebar" />
        </main>
      `,
    },
    "paragraph-component": {
      props: {
        config: { type: Object, required: true },
      },
      template: "<p>{{ config.message }}</p>",
    },
  });

  const dataResolver = new DataResolver<Configuration>();
  dataResolver
    .withResolver({
      prefix: ["context", "urlquery"],
      resolve: (context, sourcePath) => context.pageUrlQueries[sourcePath[2]],
    })
    .withResolver({
      prefix: ["context", "urlparam"],
      resolve: (context, sourcePath) => context.pageUrlParams[sourcePath[2]],
    });

  const server = new SsrServer<Configuration>();
  server.initialize({
    vueApp,
    dataResolver,
    fetchPages: () => Promise.resolve([
      {
        key: "home",
        title: "Home",
        description: "The home page of the website",
        urlPattern: "/",
        details: {
          template: "default-template",
          title: "Home",
          language: "en",
          config: { message: "Hello world!" },
          slots: {
            main: [],
            sidebar: [],
          },
        },
      },
      {
        key: "page",
        title: "Page",
        description: "Just a simple test page",
        urlPattern: "/test/:key",
        details: {
          template: "default-template",
          title: "Test",
          language: "en",
          config: { message: "Hello world!" },
          slots: {
            main: [
              {
                component: "paragraph-component",
                config: {
                  message: { $source: ["context", "urlparam", "key"] },
                },
              },
            ],
            sidebar: [
              {
                component: "paragraph-component",
                config: {
                  message: { $source: ["context", "urlquery", "message"] },
                },
              },
            ],
          },
        },
      },
    ]),
    configPath: "./config.test.json",
  });

  return server;
}

describe("SsrServer", () => {
  let server: SsrServer<Configuration>;

  beforeEach(async () => {
    server = await initializeSsrServer();
  });

  it("should work", async () => {
    const result = await supertest(server.express)
      .get("/");

    expect(result.status).toBe(200);
    expect(result.headers["content-type"]).toBe("text/html; charset=utf-8");
    expect(result.text).toMatchSnapshot();
  });

  it("should be able to fetch standard context", async () => {
    const result = await supertest(server.express)
      .get("/test/foo")
      .query({ message: "test" });

    expect(result.status).toBe(200);
    expect(result.headers["content-type"]).toBe("text/html; charset=utf-8");
    expect(result.text).toMatchSnapshot();
  });

  it("should handle static files", async () => {
    const result = await supertest(server.express)
      .get("/main.test.css");

    expect(result.status).toBe(200);
    expect(result.headers["content-type"]).toBe("text/css");
    expect(result.text).toMatchSnapshot();
  });
});
