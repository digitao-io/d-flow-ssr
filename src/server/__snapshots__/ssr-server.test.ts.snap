// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SsrServer should be able to fetch standard context 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Test</title>
  <script>window.context = {"pageDetails":{"template":"default-template","language":"en","title":"Test","config":{"message":"Hello world!"},"slots":{"main":[{"component":"paragraph-component","config":{"message":"foo"}}],"sidebar":[{"component":"paragraph-component","config":{"message":"test"}}]}}}</script>
</head>
<body>
  <div id="app"><main><h1>Hello world!</h1><p>message: Hello world!</p><!--[--><p>foo</p><!--]--><!--[--><p>test</p><!--]--></main></div>
  <script type="module" src="/src/entry-client.ts"></script>
</body>
</html>"
`;

exports[`SsrServer should handle static files 1`] = `
"body {
  margin: 0;
}"
`;

exports[`SsrServer should work 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Home</title>
  <script>window.context = {"pageDetails":{"template":"default-template","language":"en","title":"Home","config":{"message":"Hello world!"},"slots":{"main":[],"sidebar":[]}}}</script>
</head>
<body>
  <div id="app"><main><h1>Hello world!</h1><p>message: Hello world!</p><!--[--><!--]--><!--[--><!--]--></main></div>
  <script type="module" src="/src/entry-client.ts"></script>
</body>
</html>"
`;
